calyx-opt/src/passes/group_static_promotion.rs

```diff
@@ -0,0 +1,525 @@
+}
+
+impl GoDone {
+    pub fn new(ports: Vec<(ir::Id, ir::Id, u64)>) -> Self {
+        Self { ports }
+    }
+
+    /// Returns true if this is @go port
+    pub fn is_go(&self, name: &ir::Id) -> bool {
+        self.ports.iter().any(|(go, _, _)| name == go)
+    }
+
+    /// Returns true if this is a @done port
+    pub fn is_done(&self, name: &ir::Id) -> bool {
+        self.ports.iter().any(|(_, done, _)| name == done)
+    }
+
+    /// Returns the latency associated with the provided @go port if present
+    pub fn get_latency(&self, go_port: &ir::Id) -> Option<u64> {
+        self.ports.iter().find_map(|(go, _, lat)| {
+            if go == go_port {
+                Some(*lat)
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Iterate over the defined ports
+    pub fn iter(&self) -> impl Iterator<Item = &(ir::Id, ir::Id, u64)> {
+        self.ports.iter()
+    }
+}
+
+impl From<&ir::Primitive> for GoDone {
+    fn from(prim: &ir::Primitive) -> Self {
+        let done_ports: HashMap<_, _> = prim
+            .find_all_with_attr("done")
+            .map(|pd| (pd.attributes["done"], pd.name))
+            .collect();
+
+        let go_ports = prim
+            .find_all_with_attr("go")
+            .filter_map(|pd| {
+                pd.attributes.get("static").and_then(|st| {
+                    done_ports
+                        .get(&pd.attributes["go"])
+                        .map(|done_port| (pd.name, *done_port, *st))
+                })
+            })
+            .collect_vec();
+        GoDone::new(go_ports)
+    }
+}
+
+impl From<&ir::Cell> for GoDone {
+    fn from(cell: &ir::Cell) -> Self {
+        let done_ports: HashMap<_, _> = cell
+            .find_all_with_attr("done")
+            .map(|pr| {
+                let port = pr.borrow();
+                (port.attributes["done"], port.name)
+            })
+            .collect();
+
+        let go_ports = cell
+            .find_all_with_attr("go")
+            .filter_map(|pr| {
+                let port = pr.borrow();
+                port.attributes.get("static").and_then(|st| {
+                    done_ports
+                        .get(&port.attributes["go"])
+                        .map(|done_port| (port.name, *done_port, *st))
+                })
+            })
+            .collect_vec();
+        GoDone::new(go_ports)
+    }
+}
+
+/// Infer "static" annotation for groups and add "@static" annotation when
+/// (conservatively) possible.
+///
+/// Infers the number of cycles for groups where the `done`
+/// signal relies only on other `done` signals, and then inserts "static"
+/// annotations with those inferred values. If there is an existing
+/// annotation in a group that differs from an inferred value, this
+/// pass will throw an error. If a group's `done` signal relies on signals
+/// that are not only `done` signals, this pass will ignore that group.
+pub struct GroupStaticPromotion {
```

Comment:
Seems like this code is the same as `InferStaticTiming`. What is the use case for keeping two versions of this code around? If we do want to keep the two versions around, we should find some way to share the code while we keep support for `infer-static-timing` around.